### PR TITLE
ARROW-9032: [C++] Split up arrow/util/bit_util.h into multiple header files

### PR DESCRIPTION
**Pull request by wesm**
There is a lot of code in bit_util.h that is seldom used compared to certain common utilities like `BitUtil::BytesForBits`. This moves everything outside of the `BitUtil` namespace to different headers. You can see by the frequency of includes that this makes sense so that compilation units that only need some simple bit utilities are not including a lot of header code that they never use

```
$ grep -R bit_util.h ../src/ | wc -l
68
$ grep -R bitmap.h ../src/ | wc -l
7
$ grep -R bitmap_ops.h ../src/ | wc -l
15
$ grep -R bitmap_reader.h ../src/ | wc -l
9
$ grep -R bitmap_writer.h ../src/ | wc -l
6
```

This doesn't seem to affect aggregate compilation time very much but at minimum makes the code easier to navigate (in my opinion, at least). 

All the unit tests are still in bit_util_test.cc. Maybe we can improve that in a follow up patch.

### CODE REVIEW

**Comment by emkornfield** [comment]
@wesm I would appreciate if we could merge https://github.com/apache/arrow/pull/6985 first or if you prefer you take over the rebase for it?

**Comment by wesm** [comment]
@emkornfield I can definitely rebase #6985 for you

**Comment by wesm** [comment]
If this PR in principle is not objectionable would it be OK if I merged it once the build is green and then I can rebase all the PRs that it causes conflicts with?

**Comment by emkornfield** [comment]
Haven't had a chance to review it but I think it is a good thing to do, bit_util.h was definitely getting too crowded.

**Comment by wesm** [comment]
OK. I will get the build passing here and work on some dependent patches until you or someone can have a chance to review

**Comment by wesm** [comment]
Rebased

**Comment by wesm** [comment]
As far as I'm concerned this can be merged as soon as the build is green and someone signs off on it

**Comment by wesm** [comment]
The ARM failure is a flake, so this is merge-ready

File: cpp/src/arrow/util/bitmap_ops.h [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+namespace internal {
+
+// A function that visits each bit in a bitmap and calls a visitor function with a
+// boolean representation of that bit. This is intended to be analogous to
+// GenerateBits.
+template <class Visitor>
+void VisitBits(const uint8_t* bitmap, int64_t start_offset, int64_t length,
```
Comment (emkornfield) [inline cpp/src/arrow/util/bitmap_ops.h:37]: it might pay to separate VisitBis into its own header.

File: cpp/src/arrow/util/bitmap_ops.h [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+/// \brief Do a "bitmap xor" for the given bit-length on right and left
+/// buffers starting at their respective bit-offsets and put the results in
+/// out starting at the given bit offset.
+ARROW_EXPORT
+void BitmapXor(const uint8_t* left, int64_t left_offset, const uint8_t* right,
+               int64_t right_offset, int64_t length, int64_t out_offset, uint8_t* out);
+
+/// \brief Generate Bitmap with all position to `value` except for one found
+/// at `straggler_pos`.
+ARROW_EXPORT
```
Comment (emkornfield) [inline cpp/src/arrow/util/bitmap_ops.h:218]: I think it might pay to move these to a bitmap_builders.h?

File: cpp/src/arrow/util/bitset.h [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <algorithm>
```
Comment (emkornfield) [inline cpp/src/arrow/util/bitset.h:20]: might pay to name this file bitset_stack?  Not your code but is there a bifference between a bitmap and bitset in this context?

**Review by emkornfield — APPROVED** [review]
A couple of small suggestions but generally looks OK to me.

File: cpp/src/arrow/util/bitmap_ops.h (hunk 1 above)
Comment (wesm) [reply cpp/src/arrow/util/bitmap_ops.h:37]: yes, good idea

File: cpp/src/arrow/util/bitmap_ops.h (hunk 2 above)
Comment (wesm) [reply cpp/src/arrow/util/bitmap_ops.h:218]: OK

**Comment by wesm** [comment]
Will merge this once CI is passing

**Comment by wesm** [comment]
Appveyor looks OK https://ci.appveyor.com/project/wesm/arrow/builds/33339226. Merging

**Pull request closed without merging**